File: app/src/main/java/com/freestar/android/sample/recyclerview/RecyclerViewInfiniteAdsActivity.java
package com.freestar.android.sample.recyclerview;

import android.content.DialogInterface;
import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import com.freestar.android.ads.AdRequest;
import com.freestar.android.ads.AdSize;
import com.freestar.android.ads.BannerAd;
import com.freestar.android.ads.BannerAdListener;
import com.freestar.android.ads.NativeAd;
import com.freestar.android.ads.NativeAdListener;
import com.freestar.android.sample.MediationPartners;
import com.freestar.android.sample.R;
import com.google.gson.Gson;

import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;


public class RecyclerViewInfiniteAdsActivity extends AppCompatActivity implements BannerAdListener,
        NativeAdListener {

    private static final String TAG = "RecyclerView";

    private Adapter adapter;
    private LinearLayoutManager lm;
    private BannerAd bannerAd;
    private NativeAd nativeAd;
    private boolean isAdRequestInProgress;
    private int scrollDy;
    private AdRequest adRequest;
    private boolean flip;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        adRequest = new AdRequest(this);

        adapter = new Adapter(this, true);
        DataBindingUtil.setContentView(this, R.layout.activity_recyclerview);
        RecyclerView recyclerView = findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager((lm = new LinearLayoutManager(this)));
        recyclerView.setAdapter(adapter);
        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                if (newState == RecyclerView.SCROLL_STATE_IDLE || newState == RecyclerView.SCROLL_STATE_DRAGGING) {
                    Log.d(TAG, "SCROLL_STATE_IDLE");
                    int adPos = adapter.getAdPosition();
                    int vis = scrollDy >= 0 ? lm.findLastVisibleItemPosition() : lm.findFirstVisibleItemPosition();
                    if (Math.abs(vis - adPos) >= Config.AD_DISTANCE) {
                        requestAd();
                        Log.d(TAG, "REQUEST NEW AD");
                    } else {
                        Log.d(TAG, "SCROLL_STATE_IDLE BUT DON'T REQUEST NEW AD");
                    }
                }

            }

            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                Log.d(TAG, "scrolled dy: " + dy);
                /**
                 * A negative dy value means the user is scrolling in the upwards direction
                 * and a positive dy value means the user is scrolling downwards.
                 */
                scrollDy = dy;
            }
        });
        loadData();
        setTitle("RecyclerView Infinite Ads");
    }

    private void loadData() {
        new AsyncTask<Void, Void, List<Item>>() {
            @Override
            protected List<Item> doInBackground(Void... voids) {
                try {
                    return Arrays.asList(new Gson().fromJson(new InputStreamReader(RecyclerViewInfiniteAdsActivity.this.getAssets().open("media_list.json")), Item[].class));
                } catch (IOException e) {
                    e.printStackTrace();
                    return null;
                }
            }
            @Override
            protected void onPostExecute(List<Item> items) {
                ArrayList<Adapter.ItemWrapper> list = new ArrayList<>(items.size());
                for (Item item : items) {
                    list.add(new Adapter.ItemWrapper(item));
                }
                adapter.setItemWrappers(list);
                requestAd();
            }
        }.execute();
    }

    private void requestAd() {
        if (isAdRequestInProgress) {
            Log.d(TAG, "requestAd (don't do; already in progress)");
            return;
        } else {
            Log.d(TAG, "requestAd");
            isAdRequestInProgress = true;
        }
        if (!flip) {
            flip = true;
            cleanupBannerAd();
            bannerAd = new BannerAd(this);
            bannerAd.setBannerAdListener(this);
            bannerAd.setAdSize(AdSize.MEDIUM_RECTANGLE_300_250);
            bannerAd.loadAd(adRequest);
        } else {
            flip = false;
            cleanupNativeAd();
            nativeAd = new NativeAd(this);
            nativeAd.setTemplate(NativeAd.TEMPLATE_MEDIUM);
            nativeAd.setNativeAdListener(this);
            nativeAd.loadAd(adRequest);
        }
    }

    private void cleanupBannerAd() {
        if (bannerAd != null) {
            bannerAd.destroyView();
            bannerAd = null;
        }
    }
    private void cleanupNativeAd() {
        if (nativeAd != null) {
            nativeAd.destroyView();
            nativeAd = null;
        }
    }

    @Override
    public void onBannerAdLoaded(View banner, String placement) {
        Log.d(TAG, "onBannerAdLoaded");
        isAdRequestInProgress = false;
        if (banner != null) {
            int vis = scrollDy >= 0 ? lm.findLastVisibleItemPosition() : lm.findFirstVisibleItemPosition();

            int adPos = adapter.getAdPosition();
            if (adPos != -1) {
                if (Math.abs(vis - adPos) < Config.AD_DISTANCE) {
                    Log.d(TAG, "onBannerAdLoaded but there's already an ad nearby.  dont ad to list");
                    return;//there's already an ad nearby in the list
                }
            }
            vis = vis < Config.AD_DISTANCE ? Config.AD_DISTANCE : vis;
            Log.d(TAG, "onBannerAdLoaded and inserted");
            adapter.cleanupLastAd();
            adapter.insertAd(vis, (BannerAd)banner, null);
        }
    }

    @Override
    public void onBannerAdFailed(View view, String placement, int errorCode) {
        Log.d(TAG, "onBannerAdFailed: " + errorCode);
        isAdRequestInProgress = false;
    }

    @Override
    public void onBannerAdClicked(View view, String placement) {
        Log.d(TAG, "onBannerAdClicked");
    }

    @Override
    public void onBannerAdClosed(View view, String placement) {
        Log.d(TAG, "onBannerAdClosed");
        isAdRequestInProgress = false;
    }

    @Override
    public void onNativeAdLoaded(View nativeAdView, String placement) {
        Log.d(TAG, "onNativeAdLoaded");
        isAdRequestInProgress = false;
        if (nativeAdView != null) {
            int vis = scrollDy >= 0 ? lm.findLastVisibleItemPosition() : lm.findFirstVisibleItemPosition();

            int adPos = adapter.getAdPosition();
            if (adPos != -1) {
                if (Math.abs(vis - adPos) < Config.AD_DISTANCE) {
                    Log.d(TAG, "onNativeAdLoaded but there's already an ad nearby.  dont ad to list");
                    return;//there's already an ad nearby in the list
                }
            }
            vis = vis < Config.AD_DISTANCE ? Config.AD_DISTANCE : vis;
            Log.d(TAG, "onNativeAdLoaded and inserted");
            adapter.cleanupLastAd();
            adapter.insertAd(vis, null, (NativeAd)nativeAdView);
        }
    }

    @Override
    public void onNativeAdFailed(String placement, int errorCode) {
        Log.d(TAG, "onNativeAdFailed: " + errorCode);
        isAdRequestInProgress = false;
    }

    @Override
    public void onNativeAdClicked(String placement) {

    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        adapter.cleanupAll();
        cleanupNativeAd();
        cleanupBannerAd();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.choose_partners, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_choose_partners:
                choosePartners();
                return true;
            case R.id.menu_show_current_winner:
                if (bannerAd != null) {
                    new AlertDialog.Builder(this).setMessage("Last Winning Partner: " + bannerAd.getWinningPartnerName()).show();
                }
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void choosePartners() {
        MediationPartners.choosePartners(this, adRequest, MediationPartners.ADTYPE_BANNER, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                //not implemented
            }
        });
    }
}

File: Kotlin-Sample/app/src/main/java/com/freestar/sample/kotlin/RecyclerViewInfiniteAdsActivity.kt
package com.freestar.sample.kotlin

import android.os.AsyncTask
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.freestar.android.ads.*
import com.freestar.sample.kotlin.Adapter.ItemWrapper
import com.freestar.sample.kotlin.MediationPartners.choosePartners
import com.google.gson.Gson
import java.io.IOException
import java.io.InputStreamReader
import java.util.*

class RecyclerViewInfiniteAdsActivity : AppCompatActivity(), BannerAdListener, NativeAdListener {
    private var adapter: Adapter? = null
    private var lm: LinearLayoutManager? = null
    private var bannerAd: BannerAd? = null
    private var nativeAd: NativeAd? = null
    private var isAdRequestInProgress = false
    private var scrollDy = 0
    private var adRequest: AdRequest? = null
    private var flip = false
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        adRequest = AdRequest(this)
        adapter = Adapter(this, true)
        DataBindingUtil.setContentView<ViewDataBinding>(this, R.layout.activity_recyclerview)
        val recyclerView = findViewById<RecyclerView>(R.id.recyclerView)
        recyclerView.layoutManager = LinearLayoutManager(this).also { lm = it }
        recyclerView.adapter = adapter
        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                if (newState == RecyclerView.SCROLL_STATE_IDLE || newState == RecyclerView.SCROLL_STATE_DRAGGING) {
                    Log.d(TAG, "SCROLL_STATE_IDLE")
                    val adPos = adapter!!.adPosition
                    val vis =
                        if (scrollDy >= 0) lm!!.findLastVisibleItemPosition() else lm!!.findFirstVisibleItemPosition()
                    if (Math.abs(vis - adPos) >= Config.AD_DISTANCE) {
                        requestAd()
                        Log.d(TAG, "REQUEST NEW AD")
                    } else {
                        Log.d(TAG, "SCROLL_STATE_IDLE BUT DON'T REQUEST NEW AD")
                    }
                }
            }

            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                Log.d(TAG, "scrolled dy: $dy")
                /**
                 * A negative dy value means the user is scrolling in the upwards direction
                 * and a positive dy value means the user is scrolling downwards.
                 */
                scrollDy = dy
            }
        })
        loadData()
        title = "RecyclerView Infinite Ads"
    }

    private fun loadData() {
        object : AsyncTask<Void?, Void?, List<Item>?>() {
            override fun doInBackground(vararg p0: Void?): List<Item>? {
                return try {
                    Arrays.asList(
                        *Gson().fromJson(
                            InputStreamReader(
                                this@RecyclerViewInfiniteAdsActivity.assets.open(
                                    "media_list.json"
                                )
                            ), Array<Item>::class.java
                        )
                    )
                } catch (e: IOException) {
                    e.printStackTrace()
                    null
                }
            }

            override fun onPostExecute(items: List<Item>?) {
                val list = ArrayList<ItemWrapper>(
                    items!!.size
                )
                for (item in items) {
                    list.add(ItemWrapper(item))
                }
                adapter!!.setItemWrappers(list)
                requestAd()
            }

        }.execute()
    }

    private fun requestAd() {
        isAdRequestInProgress = if (isAdRequestInProgress) {
            Log.d(
                TAG,
                "requestAd (don't do; already in progress)"
            )
            return
        } else {
            Log.d(TAG, "requestAd")
            true
        }
        if (!flip) {
            flip = true
            bannerAd = BannerAd(this)
            bannerAd!!.setBannerAdListener(this)
            bannerAd!!.adSize = AdSize.MEDIUM_RECTANGLE_300_250
            bannerAd!!.loadAd(adRequest)
        } else {
            flip = false
            nativeAd = NativeAd(this)
            nativeAd!!.template = NativeAd.TEMPLATE_MEDIUM
            nativeAd!!.setNativeAdListener(this)
            nativeAd!!.loadAd(adRequest)
        }
    }

    override fun onBannerAdLoaded(banner: View, placement: String?) {
        Log.d(TAG, "onBannerAdLoaded")
        isAdRequestInProgress = false
        if (banner != null) {
            var vis =
                if (scrollDy >= 0) lm!!.findLastVisibleItemPosition() else lm!!.findFirstVisibleItemPosition()
            val adPos = adapter!!.adPosition
            if (adPos != -1) {
                if (Math.abs(vis - adPos) < Config.AD_DISTANCE) {
                    Log.d(
                        TAG,
                        "onBannerAdLoaded but there's already an ad nearby.  dont ad to list"
                    )
                    return  //there's already an ad nearby in the list
                }
            }
            vis = if (vis < Config.AD_DISTANCE) Config.AD_DISTANCE else vis
            Log.d(TAG, "onBannerAdLoaded and inserted")
            adapter!!.cleanupLastAd()
            adapter!!.insertAd(vis, banner as BannerAd, null)
        }
    }

    override fun onBannerAdFailed(view: View, placement: String?, errorCode: Int) {
        Log.d(TAG, "onBannerAdFailed: $errorCode")
        isAdRequestInProgress = false
    }

    override fun onBannerAdClicked(view: View, placement: String?) {
        Log.d(TAG, "onBannerAdClicked")
    }

    override fun onBannerAdClosed(view: View, placement: String?) {
        Log.d(TAG, "onBannerAdClosed")
        isAdRequestInProgress = false
    }

    override fun onNativeAdLoaded(nativeAdView: View, placement: String?) {
        Log.d(TAG, "onNativeAdLoaded")
        isAdRequestInProgress = false
        if (nativeAdView != null) {
            var vis =
                if (scrollDy >= 0) lm!!.findLastVisibleItemPosition() else lm!!.findFirstVisibleItemPosition()
            val adPos = adapter!!.adPosition
            if (adPos != -1) {
                if (Math.abs(vis - adPos) < Config.AD_DISTANCE) {
                    Log.d(
                        TAG,
                        "onNativeAdLoaded but there's already an ad nearby.  dont ad to list"
                    )
                    return  //there's already an ad nearby in the list
                }
            }
            vis = if (vis < Config.AD_DISTANCE) Config.AD_DISTANCE else vis
            Log.d(TAG, "onNativeAdLoaded and inserted")
            adapter!!.cleanupLastAd()
            adapter!!.insertAd(vis, null, nativeAdView as NativeAd)
        }
    }

    override fun onNativeAdFailed(placement: String?, errorCode: Int) {
        Log.d(TAG, "onNativeAdFailed: $errorCode")
        isAdRequestInProgress = false
    }

    override fun onNativeAdClicked(placement: String?) {}
    override fun onDestroy() {
        super.onDestroy()
        adapter!!.cleanupAll()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.choose_partners, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_choose_partners -> {
                choosePartners()
                true
            }
            R.id.menu_show_current_winner -> {
                if (bannerAd != null) {
                    AlertDialog.Builder(this)
                        .setMessage("Last Winning Partner: " + bannerAd!!.winningPartnerName).show()
                }
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun choosePartners() {
        choosePartners(this, adRequest!!, MediationPartners.ADTYPE_BANNER) { dialog, which ->
            //not implemented
        }
    }

    companion object {
        private const val TAG = "RecyclerView"
    }
}